File: src/main/java/ru/chernov/urlshortener/service/infrastructure/RedisService.java
package ru.chernov.urlshortener.service.infrastructure;

import org.redisson.api.RMapCache;
import org.redisson.api.RedissonClient;
import org.redisson.client.codec.Codec;
import org.springframework.stereotype.Component;
import ru.chernov.urlshortener.config.properties.RedisProperties;

import java.time.Duration;
import java.util.Optional;


@Component
public class RedisService {
    private static final String VALUE = "value";

    private final RedissonClient redissonClient;
    private final RedisProperties redisProperties;


    public RedisService(RedissonClient redissonClient, RedisProperties redisProperties) {
        this.redissonClient = redissonClient;
        this.redisProperties = redisProperties;
    }


    public <V> RMapCache<String, V> getMapCache(String key, Codec codec) {
        String redisKey = composeKey(key);
        return redissonClient.getMapCache(redisKey, codec);
    }


    public <V> Optional<V> getValue(String key, Codec codec) {
        RMapCache<String, V> mapCache = getMapCache(key, codec);
        return Optional.ofNullable(mapCache.get(VALUE));
    }


    public <V> void putValue(String key, Codec codec, V value, Duration duration) {
        var mapCache = getMapCache(key, codec);
        mapCache.fastPut(VALUE, value);
        mapCache.expire(duration);
    }


    private String composeKey(String key) {
        return redisProperties.getKeyPrefix() + key;
    }

}

File: src/main/kotlin/ru/chernov/urlshortener/service/RedisService.kt
package ru.chernov.urlshortener.service

import org.redisson.api.RMapCache
import org.redisson.api.RedissonClient
import org.redisson.client.codec.Codec
import org.springframework.stereotype.Service
import ru.chernov.urlshortener.config.properties.RedisProperties
import java.time.Duration

private const val VALUE = "value"


@Service
class RedisService(val redissonClient: RedissonClient, val redisProperties: RedisProperties) {

    fun <V> getMapCache(key: String, codec: Codec): RMapCache<String, V> {
        val redisKey: String = composeKey(key)
        return redissonClient.getMapCache(redisKey, codec)
    }


    fun <V> getValue(key: String, codec: Codec): V? {
        return getMapCache<V>(key, codec)[VALUE]
    }


    fun <V> putValue(key: String, codec: Codec, value: V, duration: Duration) {
        val mapCache = getMapCache<V>(key, codec)
        mapCache.fastPut(VALUE, value)
        mapCache.expire(duration)
    }
    

    private fun composeKey(key: String): String {
        return redisProperties.keyPrefix + key
    }

}